design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,270 @@
+shortcomings: if a user can publish clusters to a namespace with a
+`WorkloadTarget`, then they can consume any cluster. For the application owners
+that are doing self-service provisioning in this scenario, they will likely need
+to be able to create `WorkloadTarget` resources for their dynamically
+provisioned Kubernetes cluster managed resources. However, if they can do so,
+they could just as easily create a `WorkloadTarget` for the Kubernetes
+`Provider` that is only to be used for sensitive workloads. While it would be
+nice to restrict the clusters that could be published by users with the ability
+to create `WorkloadTarget` resources, there is not an immediately clear and
+elegant way to do so. It could be argued that only trusted users should have
+this permission level anyway, so it may be less of an issue than expressed here.
+Nevertheless, a potential solution is outlined [below](#future-considerations).
+
+## Technical Implementation
+
+The implementation of this proposal is relatively lightweight and could
+potentially be implemented without churn on any existing exposed APIs. The first
+step would be implementing the `WorkloadTarget` type in [core Crossplane]. It
+would not require any additional controllers.
+
+The next step would be to modify the `KubernetesApplication` [scheduling
+controller] to list `WorkloadTarget` resources in the namespace that match
+labels, instead of the current implementation that lists `KubernetesCluster`
+claims. Currently, this controller sets a reference to a `KubernetesCluster`
+claim, which is then later propagated to the child
+`KubernetesApplicationResource` objects, which use it to get a `Secret` with
+[remote cluster connection information].
+
+Because a `WorkloadTarget` may reference either a Kubernetes cluster managed
+resource or a Kubernetes `Provider` resource, both of which ultimately lead to a
+connection `Secret`, it would likely be cleanest to go ahead and get that
+`Secret` and set a reference to it on the `KubernetesApplication`, then
+propagate to the `KubernetesApplicationResource` objects. The alternative option
+would be to set a reference to either the managed resource or the `Provider`,
+then follow the references to the corresponding `Secret` in the remote cluster
+connection step in the `KubernetesApplicationResource` object controller. While
+setting the reference to the `Secret` is potentially a cross-namespace
+reference, it is one that only exists in the status of the resource, and a valid
+`WorkloadTarget` would have to exist in the `KubernetesApplication` namespace
+for that cross-namespace reference to be set by the controller.
+
+## Future Considerations
+
+One feature that would eliminate a step for self-service dynamic cluster
+provisioning and also fix the issue described above of users with permissions to
+publish *a* cluster (i.e. create `WorkloadTarget` resources) being able to
+publish *any* cluster would be to automatically create a `WorkloadTarget` as a
```

Comment:
I like this approach - perhaps the `KubernetesCluster` claim would have a `publishTarget` field to opt in, or we'd just do it inherently.

Reply:
It sounds like maybe this should be moved from a future consideration to being part of this design, what do you think?